zoneminder: 1.32.3 -> 1.34.3

danielfullmer opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Update to latest version. The version currently in nixpkgs has a lot of vulnerabilities.
See: https://github.com/ZoneMinder/zoneminder/releases

A user with an existing database would also need to run `zmupdate.pl` as the `zoneminder` user. I've added some missing perl runtime dependencies for `zmupdate.pl` as well.

I've tested this with my currently running zoneminder instance and it seems to be running correctly after the upgrade.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

CC @peterhoeg 
Closes: #54630

danielfullmer committed:
zoneminder: 1.32.3 -> 1.34.3

veprbl reviewed (dismissed):
LGTM

aanderse commented:
@danielfullmer can we run `zmupdate.pl` on every service start so the user doesn't have to? Or do we add notes somewhere mentioning this?

danielfullmer committed:
nixos/zoneminder: update on startup if needed

danielfullmer commented:
@aanderse It looks like we can run `zmupdate.pl` on startup. I originally thought it might be a problem since the script is interactive by default--but the `-nointeractive` flag works for our use. The script doesn't do anything if the database is already at the current version.

Just added a commit addressing this. I've tested upgrading from `1.32.3` to `1.34.2` in a VM and it seems to work fine.

veprbl commented:
I had my reservations about forcing an automatic schema update back when discussed in #54630. However, I'm not using zoneminder on NixOS, so I, personally, don't mind the change.

aanderse commented:
@danielfullmer sorry my bad. I forgot that automatic database schema updates aren't universally accepted among the team. This is a judgement call that you can make. Keep the schema update or not. I think I'll start a discourse thread to get some feedback on the topic in general.

danielfullmer commented:
Any commiters please hold off on merging this for the time being. I've found an issue with the UI being unresponsive when configuring cameras that I would need to debug.  (or anyone else interested)

With regard to automatic updating--I'm open to either option. If we do go for automatic schema updates I'd love to have someone else test it as well before this gets merged.  One nice feature of manually running `zmupdate.pl` interactively is that it has optional functionality to make a database backup before updating.

I'm also very busy with my thesis at the moment so I'm not sure if I will get to fixing the UI issue in the next few days.

danielfullmer committed:
zoneminder: fix timezone detection

danielfullmer committed:
zoneminder: add patch to fix improper caching

danielfullmer commented:
I've updated this PR with a few changes. It should be ready to merge after review.

Updated to [1.34.3](https://github.com/ZoneMinder/zoneminder/releases/tag/1.34.3).
Added a fix for improper timezone detection.

As far as automatic schema updates. I think we should do this only if `services.zoneminder.database.createLocally` is enabled. Zoneminder can be used with multiple instances backed by the same remote database--and we almost surely do not want to do automatic updates in that case. Conditionally updating the schema if it is a local database is also what upstream does in their [debian postinstall script](https://github.com/ZoneMinder/zoneminder/blob/c29f2cb8cfec536e444af85f71fde66ed4e58c82/distros/debian/postinst).

The UI issue was a result of zoneminder's "cache busting". They create symlinks in `/var/cache/zoneminder` to the js/css files, where the symlink has the timestamp of the destination file in the filename. This was intended to ensure that newer files would get a different URL. However, in the nix store, the files all have the timestamp "1". So when I upgraded, these symlinks would still point to the old version of zoneminder, and were not replaced. My solution was to patch the symlink filename calculation to replace the timestamp with just the hash of the zoneminder source.

veprbl commented on pkgs/servers/zoneminder/default.nix:
```diff
@@ -78,19 +78,21 @@ let
-    ./link-with-libdl.patch
+    (substituteAll {
+      src = ./0001-Don-t-use-file-timestamp-in-cache-filename.patch;
+      srcHash = src.outputHash;
```
I guess, it makes more sense to tag cache with ``` `basename $out` ``` instead of just `$src`.

danielfullmer replied:
OK. I've tested and pushed a new commit for this change.

veprbl reviewed (approved):
Looks reasonable, but I can't test this now.

veprbl commented:
@danielfullmer Please squash the last two commits and edit "zoneminder: 1.32.3 -> zoneminder 1.34.3" to "zoneminder: 1.32.3 -> 1.34.3"

danielfullmer commented:
@veprbl Just squashed those commits and fixed the commit title
